Update protection market list schema

mds1 opened the pull request:
Updates the full token list schema, repo, and tests to conform to the new Protection Market List standard. This standard can be used two ways:

1. When markets of type `ProtectionMarketInfo` do not include the `investmentOpportunity` field, the list is simply a list of deployed protection markets a user can borrow or supply to
2. When markets of type `ProtectionMarketInfo` do include the `investmentOpportunity` field, the list can be used for automating protected investment flows

The `diffTokenLists.ts` update is not done in this PR, as it doesn't seem urgent and we can move forward without that for now. But depending on how long this takes to get reviewed/merged maybe I'll do that in this PR also

Please read through the `protectionmarketlist.schema.json` file carefully and compare it against the original `tokenlist.schema.json` file. An easy way to diff the two is to copy `tokenlist.schema.json` into your checkout of this repo, right-click the two files in VSCode, and click "Compare Selected"

mds1 committed:
refactor (wip): update type definitions to reflect potential protection market list structure

lvwrence commented on src/types.ts:
```diff
@@ -24,12 +24,69 @@ export interface Tags {
+    readonly output: TokenInfo[]; // tokens the user will receive after exiting this opportunity
+    readonly rewards: TokenInfo[]; // rewards tokens the user may claim when exiting this opportunity
+  };
+  readonly signature?: 'function divest(uint256[] amounts, address recipient, uint256[] claimRewards)';
```
Shouldn't claimRewards be an array of booleans? Or I guess you're thinking of potentially withdrawing partial reward tokens?

mds1 replied:
Yep the latter. It's possible some protocols might let you specify an amount of reward tokens to claim, so figured `uint256` was useful for the extra flexibility

lvwrence commented on src/types.ts:
```diff
@@ -24,12 +24,69 @@ export interface Tags {
+}
+
+export interface Data {
+  // TODO
```
Right, so I think the only thing we need here are (and I think it may be worth moving up to `ProtectionMarketInfo` since it's just two fields:

- The HTTP endpoint to fetch current market data and position
- Some structure describing the legs of the strategy, e.g. for the saddle aleth investment opportunity it's:

```
      strategy: [
        { action: 'Borrow', destination: 'Ether', imagePath: '/tokens/eth.png' },
        { action: 'Deposit to', destination: 'Saddle', imagePath: '/platforms/saddle.png' },
        { action: 'Stake on', destination: 'Alchemix', imagePath: '/platforms/alchemix.png' },
      ],
```

although we can probably remove the `imagePath`s and turn the actions/destinations into enums

mds1 replied:
Image path seems like a good field, maybe we make it optional. Can you and @dpekar help me come up with a good list of types for the `action` enum? Feel free to just share it here and I can update the type definition

dpekar replied:
I think @tonysheng  is actually the right person to come up with a list of the action types.

Made a ticket for him here: 
https://linear.app/cozy-finance/issue/ENG-358/come-up-with-list-of-types-of-actions-leg-of-investments-can-take

dpekar replied:
Also, should we use the ids from `platformIds` (that are part of protection markets) for the destination field here?

Seems like a good idea to reuse the identifiers to make it easy for the front end.  I think imagePath being optional is good. When it's null, frontends can fall back to using the images for the ids in the  destination.

Could also add a field destinationPlatformId or something if we want to keep the destination field readable.

dpekar replied:
I think the current shape could maybe a be improved a bit.

We basically have:
1. static data about the market
2. static data about the associated invest opty
3.  dynamic data about the invest opty

3 is handled by the data endpoint, so we don't really need a data field.  

So, I'm thinking it's a little more clear like this:

```
interface ProtectionMarketInfo {
  readonly underlying: TokenInfo; // if `underlying.address === 0xEeee....EEeE`, underlying is ETH
  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
  readonly isTriggered: boolean; // true if this market was triggered

  investmentOpportunity?: {
    readonly invest: InvestScript ;
    readonly divest: DivestScript;  
    readonly dataURI: string;  
    readonly strategy: StrategyLeg[];
 }
}
```

mds1 replied:
All of this sounds good to me, I like the platform IDs idea. I think we could actually use this repo as the master copy that defines the mapping from platform ID to platform name. 

Also I like the proposed `ProtectionMarketInfo` format. Do we need to break `strategy` into an invest and divest portion? 

Do you think all of the `investmentOpportunity` fields should be required? I'm on the fence about this one

dpekar replied:
> Also I like the proposed ProtectionMarketInfo format. Do we need to break strategy into an invest and divest portion?

Nope!

> Do you think all of the investmentOpportunity fields should be required? I'm on the fence about this one

I think the min set of fields for the opty to be functional should be required and anything else is optional.  In this case, that means invest, divest & dataURI required and strategy optional.

lvwrence commented on src/types.ts:
```diff
@@ -24,12 +24,69 @@ export interface Tags {
+  readonly underlying: TokenInfo; // if `underlying.address === 0xEeee....EEeE`, underlying is ETH
+  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
+  readonly isTriggered: boolean; // true if this market was triggered
+  readonly invest?: InvestScript;
```
Shouldn't these (and divest and data) be required fields?

mds1 replied:
Hmm, this might be worth discussing. My thinking was that by making them optional you basically get two types of lists from this spec:
- Protection market lists: These do not have `invest`, `divest`, and `data` fields, and instead are just a list of money markets and protection markets in the Cozy protocol. Because you can permissionlessly create markets, this still seems useful for discovery
- Protected investment lists: These do have those three fields, and are what the cozy app would use for protected investments

dpekar replied:
Yeah if there is no invest / divest functions, it just shouldn't show up in the protected investments table.

dpekar commented on src/types.ts:
```diff
@@ -24,12 +24,69 @@ export interface Tags {
+
+export interface ProtectionMarketInfo extends TokenInfo {
+  readonly underlying: TokenInfo; // if `underlying.address === 0xEeee....EEeE`, underlying is ETH
+  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
```
Should we add an optional `imageURI` to each`ProtectionMarketInfo` item as well

mds1 replied:
`ProtectionMarketInfo` extends `TokenInfo` which already has a `logoURI` field

dpekar replied:
Oh got it.

Maybe it's conventional in smart contract dev the way you did it, but doesn't it seem a little more clear to model it this way:

```
interface ProtectionMarketInfo {
   readonly receipt: TokenInfo;
   readonly underlying: TokenInfo
 }
```

mds1 replied:
Sorry not sure I'm following—are you suggesting pulling the `receipt` token info from within the `InvestScript` and `DivestScript` types up to the root?

mds1 replied:
Now that we have the `investmentOpportunity` parent type, I think that makes sense

mds1 replied:
Like this 
![image](https://user-images.githubusercontent.com/17163988/131746196-9535a25e-fa3b-40c1-a0f3-2fa2eb5eeba1.png)

dpekar replied:
No, sorry for not being clear...

I was suggesting that the receipt token (of the cozy protection market itself) is more like a child of the market. Right now the market extends a token, so the token's fields will be flat on the market itself.

So, was suggesting removing the `extends TokenInfo` and adding an extra token field called `receipt` which has cozy market receipt token fields.

mds1 replied:
Ah, I see. Well the market itself _is_ the token, i.e. the market contract and the token contract are the same contract address. I think introducing an artificial distinction between the market and the token would be confusing

dpekar replied:
Got it makes sense :)

mds1 committed:
type definition updates

mds1 committed:
add autogenerated protectionmarketlist schema

generated with 'npx typescript-json-schema ./tsconfig.json ProtectionMarketList > src/protectionmarketlist.schema.json'

mds1 committed:
add example protection market list

mds1 committed:
replace tokenlist.schema.json with protectionmarketlist.schema.json

mds1 committed:
update README with status tracking

lvwrence commented on src/types.ts:
```diff
@@ -24,12 +24,71 @@ export interface Tags {
+// tokens and reward tokens should be sent to
+export interface DivestScript {
+  readonly address: string;
+  readonly signature?: 'function divest(uint256[] amounts, address recipient, uint256[] claimRewards)';
```
nit: does something like `receiptAmounts` and `rewardAmounts` sound better than `amounts` and `claimRewards`?

mds1 replied:
I like that, done in 87a6c5150f58e1d0d1a7864008643ead822eea46

lvwrence reviewed (approved):
LGTM, have one small suggestion about the divest signature but don't feel too strongly about it!

mds1 committed:
more explicit input names in divest function signature

mds1 committed:
update contents of JSON test files

mds1 committed:
tests pass: updated JSON schema, files, and snapshots

mds1 committed:
rename files from .tokenlist.json to .protectionmarketlist.json

mds1 committed:
cleanup

dpekar commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,441 @@
+        },
+        "logoURI": {
+          "type": "string",
+          "description": "A URI to the protection market's logo asset; if not set, interface will attempt to find a logo based on the token address; suggest SVG or PNG of size 64x64",
```
How about 128x128? 

Will be better for hi-res screens :)

mds1 replied:
Hm, yea I wonder why Uniswap chose 64x64 and 256x256 as the two options. Either way, the sizes in the descriptions are recommendations since they can't be enforced. But 256x256 sounds good to me, so updated in 42c55ff3364adc13231e658c49aaf05d36cc6430

mds1 replied:
Hm, yea I wonder why Uniswap chose 64x64 and 256x256 as the two options. Either way, the sizes in the descriptions are recommendations since they can't be enforced. But 256x256 sounds good to me, so updated in b882227e088d0a08533721a10246826abd3f3d47

dpekar commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,441 @@
+      "description": "Describes one of the steps taken to enter an investment opportunity",
+      "properties": {
+        "action": {
+          "enum": ["Borrow", "Deposit to", "Stake on"],
```
I think this is missing a couple from:
https://docs.google.com/spreadsheets/d/1L7klM8pF-7ItvzjMVj2NxkwQZQLQB7EPs1gEpy6Dftc/edit#gid=0

mds1 replied:
Oops, thanks! Fixed in 59d3b47e3afc12ef6e56862b9deb73a9739b9c07

dpekar commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,441 @@
+        },
+        "logoURI": {
+          "type": "string",
+          "description": "A URI for the logo of the strategy leg; prefer SVG or PNG of size 256x256",
```
May as well make the size of the logos match.

256x256 sound great... even better than 128x128 suggested above 😅

mds1 replied:
All should be 256x256 now! https://github.com/Cozy-Finance/protection-market-lists/pull/1#discussion_r704692017

dpekar reviewed (commented):
LGTM!  Didn't read through every example list in detail, but went through the whole spec. 

Will be interesting to see if we make any adjustments once we start with real world integration...

Thanks for the work on this @mds1 !

mds1 committed:
update logoURI recommended sizes to 256x256

mds1 committed:
add options to the StrategyLeg action enum

lvwrence commented on src/types.ts:
```diff
@@ -24,11 +24,64 @@ export interface Tags {
+// tokens and reward tokens should be sent to
+export interface DivestScript {
+  readonly address: string;
+  readonly signature?: 'function divest(uint256[] receiptAmounts, address recipient, uint256[] rewardAmounts)';
```
nit: wanna move recipient to be the first param? feels weird that it's in between the two amounts

apbendi replied:
agree it's minor but also agree recipient would "feel better" as the first arg here.

mds1 replied:
Done in 9dfee95d9da8dc8c38890621372e465b3e83d346

apbendi commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,447 @@
+    },
+    "InvestScript": {
+      "type": "object",
+      "description": "Parameters of the contract containing an `invest` method to exit an investment opportunity",
```
copy-paste typo: should say "enter" not "exit"

mds1 replied:
Ah yep, fixed in 9dfee95d9da8dc8c38890621372e465b3e83d346

apbendi commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,447 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Protection Market List",
+  "description": "Schema for lists of markets compatible with the Cozy Protocol",
```
Wonder if we should say "markets *available on* the Cozy Protocol" instead of *compatible with*. My thinking is this is slightly more accurate, since any market that would be on this list would have to be deployed *through* the protocol. "compatible with" makes it sound like something that could have been deployed separate to the protocol but is useable with it through conformity to a standard or something.

mds1 replied:
I like this, changed in 9dfee95d9da8dc8c38890621372e465b3e83d346

apbendi commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,447 @@
+          "items": {
+            "$ref": "#/definitions/TagIdentifier"
+          },
+          "maxItems": 10,
```
Just wondering why the max and why 10?

mds1 replied:
That's how Uniswap has it and I just left it since I didn't see a strong reason to change it 🤷

apbendi commented on src/protectionmarketlist.schema.json:
```diff
@@ -0,0 +1,447 @@
+      "type": "object",
+      "description": "Properties required for detailing how a user can enter a protected investment after borrowing from the protection market",
+      "properties": {
+        "dataURI": {
```
should this also have a `"format": "uri",` the same way `logoURI` does?

mds1 replied:
Good catch, added the `"format": "uri"` along with an `"example" `

apbendi commented on src/types.ts:
```diff
@@ -24,11 +24,64 @@ export interface Tags {
+}
+
+export interface StrategyLeg {
+  action: 'Borrow' | 'Swap to' | 'Deposit for' | 'Deposit to' | 'Stake on';
```
Do we feel confident this an exhaustive list? If not, what is the process like for expanding it in the future?

mds1 replied:
This is probably not exhaustive, but is what we have to start with. Expanding this would include:
1. Updating these type definitions
2. Updating the corresponding enums in `protectionmarketlist.schema.json`
3. Pushing the new `protectionmarketlist.schema.json` file to the URL we're hosting it at

apbendi reviewed (approved):
A few mostly minor comments, but overall seems great. Nice work.

mds1 committed:
schema clarifications

lvwrence commented on src/types.ts:
```diff
@@ -24,11 +24,64 @@ export interface Tags {
+
+export interface StrategyLeg {
+  action: 'Borrow' | 'Swap to' | 'Deposit for' | 'Deposit to' | 'Stake on';
+  destination: number; // Platform ID of strategy target
```
Sorry for the post-merge comment, but I feel like a string enum would work better here (e.g. "Alchemix", "Yearn"). More human-readable and intuitive
